File: 199. Binary Tree Right Side View.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def rightSideView(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if not root:
            return []
        level = 0
        queue = [(root, level)]
        res = []
        while len(queue) > 0:
            curr = queue.pop(0)
            if len(res) == curr[1]:
                res += [curr[0].val]
            if curr[0].right:
                queue += [(curr[0].right, curr[1]+1)]
            if curr[0].left:
                queue += [(curr[0].left, curr[1]+1)]
        return res

File: 199. Binary Tree Right Side View.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
#include <queue>
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        if (root == NULL) return {};
        queue <TreeNode*> nodes;
        queue <int> levels;
        nodes.push(root);
        levels.push(0);
        int level = 0;
        vector <int> res;
        while (!nodes.empty() && !levels.empty())
        {
            root = nodes.front(); nodes.pop();
            level = levels.front(); levels.pop();
            if (res.size() == level)
                res.push_back(root->val);
            if (root->right)
            {
                nodes.push(root->right);
                levels.push(level + 1);
            }
            if (root->left)
            {
                nodes.push(root->left);
                levels.push(level + 1);
            }
        }
        return res;
    }
};
